src/Compilers/CSharp/Portable/Binder/Lifetime.cs

```diff
@@ -0,0 +1,129 @@
+{
+    /// <summary>
+    /// A representation of the program region in which a 'ref' can be used,
+    /// which is limited to what is expressible in C#.
+    /// </summary>
+    /// <remarks>
+    /// For example, in this design, all lifetimes have a known relationship to all other lifetimes.
+    /// </remarks>
+    internal readonly struct Lifetime
+    {
+        private const uint CallingMethodRaw = 0;
+        private const uint ReturnOnlyRaw = 1;
+        private const uint CurrentMethodRaw = 2;
+
+        /// <summary>
+        /// For the purpose of escape verification we operate with the depth of local scopes.
+        /// The depth is a uint, with smaller number representing shallower/wider scopes.
+        /// Since sibling scopes do not intersect and a value cannot escape from one to another without
+        /// escaping to a wider scope, we can use simple depth numbering without ambiguity.
+        /// </summary>
+        private readonly uint _value;
+        private Lifetime(uint value) => _value = value;
+
+        /// <summary>
+        /// The "calling method" scope that is outside of the containing method/lambda.
+        /// If something can escape to this scope, it can escape to any scope in a given method through a ref parameter or return.
+        /// </summary>
+        public static Lifetime CallingMethod => new Lifetime(CallingMethodRaw);
+
+        /// <summary>
+        /// The "return-only" scope that is outside of the containing method/lambda.
+        /// If something can escape to this scope, it can escape to any scope in a given method or can be returned, but it can't escape through a ref parameter.
+        /// </summary>
+        public static Lifetime ReturnOnly => new Lifetime(ReturnOnlyRaw);
+
+        /// <summary>
+        /// The "current method" scope that is just inside the containing method/lambda.
+        /// If something can escape to this scope, it can escape to any scope in a given method, but cannot be returned.
+        /// </summary>
+        public static Lifetime CurrentMethod => new Lifetime(CurrentMethodRaw);
+
+        /// <summary>
+        /// Gets a lifetime which is "empty". i.e. which refers to a variable whose storage is never allocated.
+        /// </summary>
+        public static Lifetime Empty => new Lifetime(uint.MaxValue);
+
+        /// <summary>
+        /// Gets a lifetime which is narrower than the given lifetime.
+        /// Used to "enter" a nested local scope.
+        /// </summary>
+        public Lifetime Narrower()
+        {
+            var result = new Lifetime(this._value + 1);
+            // Narrower() operator should always result in a local lifetime
+            Debug.Assert(!result.IsReturnable);
+            return result;
+        }
+
+        /// <summary>
+        /// Gets a lifetime which is wider than the given lifetime.
+        /// Used to "exit" a nested local scope.
+        /// </summary>
+        public Lifetime Wider()
+        {
```

Comment:
```suggestion
        {
            Debug.Assert(this._value > 0);
```

Reply:
`!IsReturnable` already implies this, but I can add this assertion also if you feel it improves clarity.

Reply:
Have rewritten these asserts.